client/webserver/webserver.go

```diff
@@ -0,0 +1,401 @@
+			log.Errorf("no HTML template files found")
+			return
+		}
+	}
+
+	// Prepare the templates.
+	bb := "bodybuilder"
+	tmpl := newTemplates(fp(root, "src/html"), reloadHTML).
+		addTemplate("login", bb).
+		addTemplate("register", bb).
+		addTemplate("markets", bb).
+		addTemplate("wallets", bb).
+		addTemplate("settings", bb)
+	err := tmpl.buildErr()
+	if err != nil {
+		log.Errorf(err.Error())
+		return
+	}
+
+	// Create an HTTP router.
+	mux := chi.NewRouter()
+	httpServer := &http.Server{
+		Handler:      mux,
+		ReadTimeout:  rpcTimeoutSeconds * time.Second, // slow requests should not hold connections opened
+		WriteTimeout: rpcTimeoutSeconds * time.Second, // hung responses must die
+	}
+
+	// Make the server here so its methods can be registered.
+	s := &webServer{
+		ctx:  ctx,
+		core: core,
+		html: tmpl,
+	}
+
+	// Middleware
+	mux.Use(middleware.Recoverer)
+	mux.Use(s.authMiddleware)
+	// Websocket endpoint
+	mux.Get("/ws", s.handleWS)
+	// Webpages
+	mux.Get("/", s.handleHome)
+	mux.Get("/register", s.handleRegister)
+	mux.Get("/login", s.handleLogin)
+	mux.Get("/markets", s.handleMarkets)
+	mux.Get("/wallets", s.handleWallets)
+	mux.Get("/settings", s.handleSettings)
+	mux.Route("/api", func(r chi.Router) {
+		r.Use(middleware.AllowContentType("application/json"))
+		r.Post("/register", s.apiRegister)
+		r.Post("/login", s.apiLogin)
+	})
+	// Files
+	fileServer(mux, "/js", fp(root, "dist"))
+	fileServer(mux, "/css", fp(root, "dist"))
+	fileServer(mux, "/img", fp(root, "src/img"))
+	fileServer(mux, "/font", fp(root, "src/font"))
+
+	// Start serving.
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Errorf("Can't listen on %s. web server quitting: %v", addr, err)
+		return
+	}
+	// Close the listener on context cancellation.
+	go func() {
+		<-ctx.Done()
+		err := listener.Close()
+		if err != nil {
+			log.Errorf("Problem shutting down rpc: %v", err)
+		}
+	}()
+	log.Infof("Web server listening on %s", listener.Addr())
+	err = httpServer.Serve(listener)
+	if err != http.ErrServerClosed {
+		log.Warnf("unexpected (http.Server).Serve error: %v", err)
+	}
+	log.Infof("Web server off")
+}
+
+// auth creates, stores, and returns a new auth token.
+func (s *webServer) auth() string {
+	// Create a token to identify the user. Only one token can be active at
+	// a time = only 1 authorized device at a time.
+	b := make([]byte, 32)
+	rand.Read(b)
+	token := hex.EncodeToString(b)
+	s.mtx.Lock()
+	s.authToken = token
+	s.mtx.Unlock()
+	return token
+}
+
+// token returns the current auth token.
+func (s *webServer) token() string {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	return s.authToken
+}
+
+// watchMarket watches the specified market, cancelling any other market
+// subscriptions.
+func (s *webServer) watchMarket(dex, mkt string) (*core.OrderBook, error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	old := s.openMarket
+	if old != nil {
+		log.Tracef("ending monitoring of market %s", old.market)
+		old.kill()
+	}
```

Comment:
I expect users would want to have more than one browser tab/window open with different markets.  Does this prevent that?